File: app/src/main/java/social/entourage/android/view/EntourageTitleView.java
package social.entourage.android.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.View;
import android.widget.ImageButton;
import android.widget.RelativeLayout;
import android.widget.TextView;

import social.entourage.android.R;

/**
 * Custom title view with a close button and a title
 */
public class EntourageTitleView extends RelativeLayout {

    private TextView titleTextView;
    private ImageButton closeButton;
    private TextView actionButton;

    private String mTitle;
    private Drawable mCloseButtonDrawable;

    public EntourageTitleView(Context context) {
        super(context);
        init(null, 0);
    }

    public EntourageTitleView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs, 0);
    }

    public EntourageTitleView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(attrs, defStyle);
    }

    private void init(AttributeSet attrs, int defStyle) {
        inflate(getContext(), R.layout.layout_view_title, this);

        titleTextView = findViewById(R.id.title_text);
        closeButton = findViewById(R.id.title_close_button);
        actionButton = findViewById(R.id.title_action_button);
        View separator = findViewById(R.id.title_separator);

        // Load attributes
        final TypedArray a = getContext().obtainStyledAttributes(
                attrs, R.styleable.EntourageTitleView, defStyle, 0);

        mTitle = a.getString(
                R.styleable.EntourageTitleView_entourageTitle);
        titleTextView.setText(mTitle);

        if (a.hasValue(R.styleable.EntourageTitleView_android_textColor)) {
            titleTextView.setTextColor(a.getColor(R.styleable.EntourageTitleView_android_textColor, getResources().getColor(R.color.greyish_brown)));
        }

        if (a.hasValue(R.styleable.EntourageTitleView_entourageTitleCloseDrawable)) {
            mCloseButtonDrawable = a.getDrawable(
                    R.styleable.EntourageTitleView_entourageTitleCloseDrawable);
            closeButton.setImageDrawable(mCloseButtonDrawable);
        }

        actionButton.setText(a.getString(R.styleable.EntourageTitleView_entourageTitleAction));

        setBackgroundResource(a.getResourceId(R.styleable.EntourageTitleView_android_background, R.color.background));

        boolean showSeparator = a.getBoolean(R.styleable.EntourageTitleView_entourageShowSeparator, true);
        separator.setVisibility(showSeparator ? VISIBLE : GONE);

        a.recycle();
    }

    public void setTitle(final String mTitle) {
        this.mTitle = mTitle;
        titleTextView.setText(mTitle);
    }
}

File: app/src/main/java/social/entourage/android/view/EntourageTitleView.kt
package social.entourage.android.view

import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.view.View
import android.widget.RelativeLayout
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.DrawableCompat
import kotlinx.android.synthetic.main.layout_view_title.view.*
import social.entourage.android.R

/**
 * Custom title view with a close button and a title
 */
class EntourageTitleView : RelativeLayout {

    constructor(context: Context?) : super(context) {
        init(null, 0)
    }

    constructor(context: Context?, attrs: AttributeSet) : super(context, attrs) {
        init(attrs, 0)
    }

    constructor(context: Context?, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        init(attrs, defStyle)
    }

    private fun init(attrs: AttributeSet?, defStyle: Int) {
        View.inflate(context, R.layout.layout_view_title, this)
        // Load attributes
        val attributes = context.obtainStyledAttributes(
                attrs, R.styleable.EntourageTitleView, defStyle, 0)
        title_text?.text = attributes.getString(R.styleable.EntourageTitleView_entourageTitle)
        if (attributes.hasValue(R.styleable.EntourageTitleView_android_textColor)) {
            title_text?.setTextColor(attributes.getColor(R.styleable.EntourageTitleView_android_textColor, ContextCompat.getColor(context, R.color.greyish_brown)))
        }
        if (attributes.hasValue(R.styleable.EntourageTitleView_entourageTitleCloseDrawable)) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                title_close_button?.setImageDrawable(attributes.getDrawable(
                    R.styleable.EntourageTitleView_entourageTitleCloseDrawable))
            } else {
                title_close_button?.setImageDrawable(ContextCompat.getDrawable(context,attributes.getResourceId(
                        R.styleable.EntourageTitleView_entourageTitleCloseDrawable, 0)))
            }

        }
        if (attributes.hasValue(R.styleable.EntourageTitleView_entourageTitleCloseDrawableTint)) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                title_close_button?.imageTintList = attributes.getColorStateList(
                        R.styleable.EntourageTitleView_entourageTitleCloseDrawableTint)
            } else {
                ContextCompat.getDrawable(context,attributes.getResourceId(R.styleable.EntourageTitleView_entourageTitleCloseDrawable, 0))?.let {
                    DrawableCompat.setTint(it, attributes.getColor(R.styleable.EntourageTitleView_entourageTitleCloseDrawableTint, ContextCompat.getColor(context, R.color.accent)))
                }
            }
        }
        title_action_button?.text = attributes.getString(R.styleable.EntourageTitleView_entourageTitleAction)
        setBackgroundResource(attributes.getResourceId(R.styleable.EntourageTitleView_android_background, R.color.background))
        title_separator?.visibility = if (attributes.getBoolean(R.styleable.EntourageTitleView_entourageShowSeparator, true)) View.VISIBLE else View.GONE
        attributes.recycle()
    }

    fun setTitle(title: String) {
        title_text?.text = title
    }
}